### PR TITLE
Adds test of Scylla Manager Alternator restore

### LINKED ISSUE #11696: Scylla Manager feature test for Alternator backup and restore
Alternator requires separate (non-CQL based) workflow for backup and restore - which was already discussed in several places - https://github.com/scylladb/scylla-manager/issues/4509
The new workflow should also have an E2E test, as requested here: https://github.com/scylladb/scylla-manager/issues/4505

If I understand correctly, the existing `mgmt_cli_test.ManagerRestoreTests.test_restore_backup_with_task`  https://github.com/scylladb/scylla-cluster-tests/blob/ac10fd8c948f3d4b2a842f4699f19744e77294b1/mgmt_cli_test.py#L832
can be adapted to support Alternator:
- `create_backup_task` and `create_restore_task` may need to add some arguments to manager command, (@Michal-Leszczynski or is it going to automatically recognize Alternator tables?)
https://github.com/scylladb/scylla-cluster-tests/blob/ac10fd8c948f3d4b2a842f4699f19744e77294b1/sdcm/mgmt/cli.py#L605
- the workload of the test may be defined by setting different loader in `prepare_write_cmd` (to setup tables) and `stress_read_cmd` (to check expected state of tables):
https://github.com/scylladb/scylla-cluster-tests/blob/ac10fd8c948f3d4b2a842f4699f19744e77294b1/test-cases/manager/manager-backup-restore-set-dataset.yaml#L6-L7

@mikliapko, is that correct? Are there any nuances I should be aware of? 

Regarding loaders, ycsb is currently the only loader that supports Alternator, right? It may be used directly, but I am not sure if it supports some corner cases we would want to test:
- specific characters in names (especially dashes and dots)
- GSI with two key columns and computed columns
- LSI

So this may need extending the loader.

Is it worth trying to run whole suites or would this test be enough to test Alternator support? @ScyllaPiotr @Michal-Leszczynski @mikliapko

### PR DESCRIPTION
**Pull request by m-szymon**
This patch adds `test_alternator_backup_feature` which checks Scylla Manager's feature
for backing up and restoring Alternator data.
In general it copies workflow of previous CQL backup test `test_restore_backup_with_task`, but:
- it pre-creates Alternator table,
- it fills table and runs post-restore check using YSCB,
- table is deleted between backup and restore

Besides plain table+data restore, it checks additional (configurable with `alternator_test_table` option) features:
- restoring table tags
- table name including special characters
- restoring table with secondary indexes

Restored index content could be tested with:
    - YCSB updated to support index in read command (https://github.com/scylladb/YCSB/pull/20)
    - stress commands in `test-cases/manager/manager-regression-alternator-singleDC-set-distro.yaml`.
Before running stress commands, test waits for the indexes to achieve expected number of records.

When defined in configuration, the tests passes to Scylla Manager additional command line parameters
 `--alternator-access-key-id` `--alternator-secret-access-key`.

Fixes #11696

### Testing

scylla 2025.1.7: https://argus.scylladb.com/tests/scylla-cluster-tests/60ebecb9-6891-40b8-be51-0f41832a5d6a
scylla 2025.2.2: https://argus.scylladb.com/tests/scylla-cluster-tests/d52d95d8-4626-4493-a1c9-36a7eac3783b
scylla 2025.3.1: https://argus.scylladb.com/tests/scylla-cluster-tests/4cff2c73-54e8-4810-b014-b497b48c2e70
scylla 2025.4.0: https://argus.scylladb.com/tests/scylla-cluster-tests/aa26fb8f-cf9d-42fd-83b4-44eee7f6bc57
scylla master:latest: https://argus.scylladb.com/tests/scylla-cluster-tests/35a7b70d-33c6-40d9-ae63-18bc3370888a

previous runs without secondary index YSCB patch:
- scylla master:latest https://argus.scylladb.com/tests/scylla-cluster-tests/f16b44d7-7ce6-4fee-bcc6-cedb11c837c7
- scylla 2025.3.1 https://argus.scylladb.com/tests/scylla-cluster-tests/3df93354-590d-4d53-b998-4950af699e71
- scylla 2025.2.2 https://argus.scylladb.com/tests/scylla-cluster-tests/d10d6c6b-85eb-4d85-9d49-f26b60136cd1
- scylla 2025.1.7 https://argus.scylladb.com/tests/scylla-cluster-tests/7f516714-1e98-4b9c-8782-2cc1d80f7924

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

**Review by mikliapko — COMMENTED** [review]
Did a high-level overview of suggested test scenario with major focus on logical part since the PR is still in Draft.

**Comment by m-szymon** [comment]
> @m-szymon Do you have any test runs for designed test? Could you please attach them to the run?

I added link to cover letter (https://argus.scylladb.com/tests/scylla-cluster-tests/7cc93e37-c85f-4162-9a8f-a5a1b8cbfb4e).

@Michal-Leszczynski, @radoslawcybulski, could you also have a look if this makes sense on high level?

File: mgmt_cli_test.py [hunk 1]
```diff
@@ -693,6 +693,15 @@ def verify_backup_success(self, mgr_cluster, backup_task, ks_names: list = None,
             self.restore_backup_without_manager(mgr_cluster=mgr_cluster, snapshot_tag=snapshot_tag,
                                                 ks_tables_list=ks_tables_map)
 
+    def verify_alternator_backup_success(self, mgr_cluster, backup_task, delete_tables: list = None, timeout=None):
+        for table_name in delete_tables:
+            self.log.info(f'running delete on {table_name}')
+            self.alternator.delete_table(self.db_cluster.nodes[0], table_name=table_name, wait_until_table_removed=True)
+        self.restore_backup_with_task(mgr_cluster=mgr_cluster, snapshot_tag=backup_task.get_snapshot_tag(),
+                                      timeout=timeout, restore_schema=True)
```
[the remark this answers is not in the document]
Comment (Michal-Leszczynski) [reply mgmt_cli_test.py:701]: As SM basically just applies the backed up output of `DESCRIBE SCHEMA WITH INTERNALS`, it has to be able to create all entities listed there - meaning all user keyspaces in the backed up schema should be deleted before restoring the schema with SM.

**Comment by Michal-Leszczynski** [comment]
With the [mentioned](https://github.com/scylladb/scylla-cluster-tests/pull/11786#discussion_r2352183361) change the test (with special chars in table name and uncommented features except the one blocked by YCBS PR) [passed](https://jenkins.scylladb.com/job/scylla-staging/job/michal-leszczynski/job/SCT-ubuntu24-manager-backup-alternator/8/) against SM build from this [PR](https://github.com/scylladb/scylla-manager/pull/4569).

**Comment by Michal-Leszczynski** [comment]
@m-szymon what's the ETA for this PR to me merged (also with the support for validating GSI/LSI restoration)?
All alternator backup/restore implementation on SM side will be available on SM master in a few hours, so it looks like the SCT test will be the last one before we can release SM with alternator backup/restore support.

**Comment by m-szymon** [comment]
@Michal-Leszczynski 
I get an error `Error: unknown flag: --alternator-access-key-id` here
https://argus.scylladb.com/tests/scylla-cluster-tests/6541e43e-bf4a-4a86-a35a-e16b2aed4ec8/events,
and I can't figure out what I am doing wrong. When I run it locally, it works. 

> we could:
> 
>     * restore schema and data
> 
>     * validate restored table and GSI
> 
>     * sleep for X time
> 
>     * validate restored table LSI
> 
> 
> With a high enough sleep duration it shouldn't be flaky and we should have some coverage over LSI restoration.

Schema is validated, table data are checked.
Enabling GSI and LSI data validation - with YSCB update - deserves separate PR.

> Do you mean it's not a blocker for this PR or for releasing alternator backup/restore support in SM?

I think it shouldn't block anything.

**Review by Michal-Leszczynski — APPROVED** [review]

File: jenkins-pipelines/manager/ubuntu24-manager-backup-alternator.jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+managerPipeline(
+    backend: 'aws',
+    region: 'us-east-1',
+    test_name: 'mgmt_cli_test.ManagerBackupTests.test_alternator_backup_feature',
+    test_config: 'test-cases/manager/manager-regression-alternator-singleDC-set-distro.yaml',
+
+    post_behavior_db_nodes: 'destroy',
+    post_behavior_loader_nodes: 'destroy',
+    post_behavior_monitor_nodes: 'destroy',
```
Comment (mikliapko) [inline jenkins-pipelines/manager/ubuntu24-manager-backup-alternator.jenkinsfile:14]: These parameters are not required to be set explicitly in jenkinsfile anymore and set by default.
Check out https://github.com/scylladb/scylla-cluster-tests/blob/master/defaults/test_default.yaml#L51

You will see them in all Manager related jenkinsfiles as they was not refactored yet.

File: sdcm/mgmt/operations.py [hunk 3]
```diff
@@ -700,6 +700,15 @@ def verify_backup_success(self, mgr_cluster, backup_task, ks_names: list = None,
...
                                                 ks_tables_list=ks_tables_map)
 
+    def verify_alternator_backup_success(self, mgr_cluster, backup_task, delete_tables: list = None, timeout=None):
+        for table_name in delete_tables:
+            self.log.info(f'running delete on {table_name}')
+            self.alternator.delete_table(self.db_cluster.nodes[0], table_name=table_name, wait_until_table_removed=True)
+        self.restore_backup_with_task(mgr_cluster=mgr_cluster, snapshot_tag=backup_task.get_snapshot_tag(),
+                                      timeout=timeout, restore_schema=True)
+        self.restore_backup_with_task(mgr_cluster=mgr_cluster, snapshot_tag=backup_task.get_snapshot_tag(),
+                                      timeout=timeout, restore_data=True)
```
Comment (mikliapko) [inline sdcm/mgmt/operations.py:710]: Shouldn't we run some read operations afterwards to verify the restore from backup success?

File: sdcm/tester.py [hunk 4]
```diff
@@ -1271,6 +1268,41 @@ def pre_create_alternator_tables(self):
                 self.alternator.update_table_ttl(node=node, table_name=alternator.consts.TABLE_NAME)
                 self.alternator.update_table_ttl(node=node, table_name=alternator.consts.NO_LWT_TABLE_NAME)
 
+    def pre_create_alternator_backuped_tables(self, **kwargs):
```
Comment (soyacz) [inline sdcm/tester.py:1271]: I'd prefer not adding more to `tester.py` file.
I propose to add `node`, `params` and `alternator` parameters so this method could be moved to `sdcm.utils.alternator` module.
(same for methods/ctx mgr below)

**Comment by Michal-Leszczynski** [comment]
@m-szymon I see that the work on this and YCBS PR is still in progress. I understand that merging them will require some additional rounds of reviews etc., but would it be possible to test SM against this test using this branch and custom YCBS branch? I wouldn't mind releasing alternator backup/restore support in SM if we executed such test - it would be fine to merge those PRs later on.

I'm writing this because we will need to release SM compatible with scylla 2025.4.0, and it would be nice to include the alternator backup/restore functionality in it instead of trying to out cherry-pick those commits and be forced to make another minor release shortly afterwards.

cc: @ScyllaPiotr @karol-kokoszka

**Comment by Michal-Leszczynski** [comment]
> I have been running this test with custom YCSB branch locally - I don't know if I can do it in Argus - and it was passing.

If this test tested all interesting features (e.g. problematic names, LSI and GSI contents), then it should be good enough.

> In the late evening I should be ready.

It's not that urgent that we need it for today/tomorrow, but it would be good to have the results this week:)
(unless we already have everything we need from local execution)

**Comment by m-szymon** [comment]
@Michal-Leszczynski YSCB patch is merged.
@fruch Can you build official image of YSCB, so that I can use it here?

These are runs with my image:
scylla 2025.1.7: https://argus.scylladb.com/tests/scylla-cluster-tests/60ebecb9-6891-40b8-be51-0f41832a5d6a
scylla 2025.2.2: https://argus.scylladb.com/tests/scylla-cluster-tests/d52d95d8-4626-4493-a1c9-36a7eac3783b
scylla 2025.3.1: https://argus.scylladb.com/tests/scylla-cluster-tests/4cff2c73-54e8-4810-b014-b497b48c2e70
scylla 2025.4.0: https://argus.scylladb.com/tests/scylla-cluster-tests/aa26fb8f-cf9d-42fd-83b4-44eee7f6bc57
scylla master:latest: https://argus.scylladb.com/tests/scylla-cluster-tests/35a7b70d-33c6-40d9-ae63-18bc3370888a

2025.4.0 failed, but it is scylla that crashed at the end :shrug:

**Comment by soyacz** [comment]
@CodeLieutenant can you build YCSB?

**Review by soyacz — APPROVED** [review]
From my perspective looks ok, but someone is needed to confirm the test flow is right and comments needs to be resolved.
Also consider setting backport label for 2025.4 (if ought to be run there)
